### PR TITLE
Support eviction API for VMIs with eviction strategy

### PR DESCRIPTION
**Pull request by danielBelenky**
This PR adds support for the eviction API and triggers a migration for
VMIs with an eviction strategy when they are being evicted. This allows
users to properly use node drain to evacuate an entire node and allows
the descheduler to do it's work as intended and to move VMIs from nodes
that are under some kind of pressure.

In order to trigger a migration when an eviction is being requested on a
virt launcher pod, we intercept the eviction request and mark the target
pod. The evacuation controller then kicks in and the migration from there.

The migration tests were not stable because we could potentially drain the
single master node that we have in the cluster so I've also modified those
tests to ensure that the initial VMI is created on the compute node. 

**Release note**:

```release-note
Added support for the eviction API for VMIs with eviction strategy. This enables VMIs to be live-migrated when the node is drained or when the descheduler wants to move a VMI to a different node.
```

### CODE REVIEW

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (rmohr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:51]: On a quick check it looks like the conditions is never removed, can that be?
I think that an explicit status subsection where you can log the eviction timepoint and the node on which it happens is better than a condition. I would expect that the status subsection is cleared if the VMI then runs on another node.

Comment (danielBelenky) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:51]: Heh :) I just thought on this situation a few minutes ago too.
Thanks, I'll fix that.

File: tests/migration_test.go
Comment (danielBelenky) [inline tests/migration_test.go:1495]: Working on this one.

File: pkg/virt-operator/creation/components/webhooks.go
Comment (enp0s3) [inline pkg/virt-operator/creation/components/webhooks.go:194]: @danielBelenky remove this one

**Review by enp0s3 — CHANGES_REQUESTED** [review]
Awesome!
Found a few things while doing a first iteration.

**Review by enp0s3 — CHANGES_REQUESTED** [review]
A few more points

File: pkg/virt-operator/creation/components/webhooks.go
Comment (davidvossel) [inline pkg/virt-operator/creation/components/webhooks.go:237]: shouldn't we be using ignore on failure here? It looks like the failure policy it to fail

Like your comment points out, this will mean that all evictions are blocked across the cluster if our component is down.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go
Comment (ashleyschuett) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go:99]: lets use the `NewFakeClusterConfigUsingKV` instead so that this doesn't need to be updated in the near future

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go
Comment (ashleyschuett) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter_test.go:341]: same as above

**Comment by davidvossel** [comment]
> [rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system] VM Live Migration with a live-migrate eviction strategy set [ref_id:2293] with a VMI running with an eviction strategy set with node tainted during node drain [test_id:2221] should migrate a VMI under load to another node expand_more

func test failure.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (davidvossel) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:62]: I'd expect the eviction to be retried by whatever is calling this if the POST request returns an error.  The client would have to be resilient to errors like this because they can already occur today if connectivity issues are encountered while issuing the evictions.

It would be safer for us to reject the eviction here and trust the client issuing the eviction to handle that. 

Our contract with the user when they set evictionStrategy to liveMigrate is that we're not going to stop/start the VM, but block until migration occurs.

Comment (danielBelenky) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go]: As we've discussed offline last week, I've moved this logic to a mutating webhook to ensure that we are being called. If this step will fail, the eviction will be blocked due to the PDB as expected.

**Comment by danielBelenky** [comment]
/hold

Making some adjustments. Unit tests are expected to fail at this point, I'm stabilizing the system tests.

**Comment by danielBelenky** [comment]
/hold cancel

this is now ready for code-review

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:73]: Any particular reason to use this code?

File: pkg/virt-controller/watch/drain/evacuation/evacuation.go
Comment (xpivarc) [inline pkg/virt-controller/watch/drain/evacuation/evacuation.go:335]: This is still true or?

File: pkg/virt-controller/watch/drain/evacuation/evacuation.go
Comment (xpivarc) [inline pkg/virt-controller/watch/drain/evacuation/evacuation.go:436]: Is there a reason for a copy?

File: pkg/virt-controller/watch/drain/evacuation/evacuation.go
Comment (xpivarc) [inline pkg/virt-controller/watch/drain/evacuation/evacuation.go:323]: This function name mislead me. Can I suggest to rename it to something like `VMIsToEvict`.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (dhiller) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:42]: I've been trying to wrap my head around this for some time now. What this code is telling me from reading is that

*if I shouldn't try to deny the eviction\* then I **should deny** the eviction*

\* aka negating the method return value from `shouldTryInterceptEviction`

I feel like this is an error in the naming of the methods somehow? Can you please elaborate on this?

**Review by dhiller — COMMENTED** [review]
Looks very good already, just a question for now.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:87]: Is this used?

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:49]: We propably shoud deny this but should we also suggest retry?

File: staging/src/kubevirt.io/client-go/api/v1/types.go
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:211]: I was just extra secure here... if we had two conditions with VirtualMachineInstanceIsMigratable type one being false and one true... This should not happen.

**Review by xpivarc — COMMENTED** [review]
One last thing.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (danielBelenky) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:87]: no :) I forgot to remove it.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (danielBelenky) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:49]: It is not just for a retry.. it is to block the eviction process on the client's side until it completes. It is up to the client to decide what to do at this point. I agree that this "contract" with the client is not perfect here... it feels like we're missing some communication and error channels at this point, but that's what we have in our tool box right now. We might consider improving this in the future on `kubelet`.

Comment (xpivarc) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:49]: Maybe I misunderstand here something... but retry in this case would not make difference, right? So we should probably deny this but with a different response code so clients will not automaticly retry and the user will need to manual interfere.

Comment (danielBelenky) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:49]: By "retry" I mean - this is our only way to block the client from proceeding and tell him that something is wrong and he should take a look at his VMI. He will then see the condition that his VMI is not migratable.

**Review by dhiller — APPROVED** [review]
/lgtm
/approve

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (rmohr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:48]: Indeed. We definitely want to handle this better, but this is right now the only option which we have with k8s.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (davidvossel) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:68]: This needs to be a Patch operation. virt-api doesn't own a running VMI, virt-handler does.

See how virt-controller's vmi controller handles Patching running virtual machines. The risk here is higher for a update failure when patch isn't used.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (davidvossel) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:40]: We really need to be using an informer here rather than making GET calls within the api webhook. This goes for both the VMI and POD GET requests.

in your 

```
type PodEvictionAdmitter struct {
	ClusterConfig *virtconfig.ClusterConfig
	VirtClient    kubecli.KubevirtClient
}
```

struct, that ClusterConfig object contains informers. You simply need to add a vmi and vmi pod informer there which will be immediately accessible to this webhook code. That's the pattern we've already established for other webhooks.

File: pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go
Comment (davidvossel) [inline pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:17]: add vmi and pod informers to the ClusterConfig

**Comment by davidvossel** [comment]
@danielBelenky Sorry for the late review. I missed that this PR was related to the webhook logic. we should use informers in our webhooks and avoid calling Update directly.

The logic you have now will probably work okay most of the time, these suggestions will improve the resiliency though. 

/cc @dhiller @xpivarc

**Pull request merged**
